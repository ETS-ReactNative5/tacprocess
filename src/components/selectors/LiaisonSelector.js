import React from 'react'
import propTypes from 'prop-types'
import {getSaltAstronomerName, getSaltAstronomerUsername, hasLiaison} from '../../util/salt-astronomer'
import {isLiaisonAstronomerUpdated} from '../../util/proposal-filtering'

const getEventValue = (event, astronomers) => {
  event.preventDefault()
  if (event.target.name === 'selector') {
    if (event.target.value === 'none' ) return null
    return getSaltAstronomerUsername(event.target.value, astronomers)
  }
  if (event.target.checked) return event.target.value
  return null
}

const liaisonChange = (event, astronomers, proposalCode, setLiaison, send ) => {
	setLiaison(getEventValue(event, astronomers), proposalCode)
  send(getEventValue(event, astronomers), proposalCode)
}

const LiaisonSelector = ({proposal, astronomers, setLiaison, initProposals, username, canAssign, send}) => {
  const col = isLiaisonAstronomerUpdated(proposal, initProposals) ? {color: 'blue'} : {color: 'black'}
  const liaison = (proposal.possibleLiaison) ? getSaltAstronomerName(proposal.possibleLiaison, astronomers) :
    getSaltAstronomerName(proposal.liaisonAstronomer, astronomers)
  return (
    <td>
      {
        canAssign ?
          <select className='setLiaison' style={ col } defaultValue={ liaison } onChange={
            e => liaisonChange(e, astronomers, proposal.proposalCode, setLiaison, send)
					} name='selector'>
            {
              !hasLiaison(proposal.proposalCode, initProposals) && <option value='none'>none</option>
            }
            {
              astronomers.map(op => (
                  <option key={ op.username } value={ op.name } name={ op.username }>
                    {op.name}
                  </option>
              ))
            }
          </select> :
          <div>
            <input
              className='setLiaison'
              checked={ proposal.liaisonAstronomer }
              disabled={ hasLiaison(proposal.proposalCode, initProposals) }
              type='checkbox'
              value={ username }
              onChange={ e => liaisonChange(e, astronomers, proposal.proposalCode, setLiaison, send) }/>
            {
              proposal.liaisonAstronomer ?
              <a style={ col }> {liaison} </a>: <a style={ {color: 'red'} }>Select</a>
            }
          </div>
      }
    </td>
)}

LiaisonSelector.propTypes = {
  proposal: propTypes.object.isRequired,
  astronomers: propTypes.array.isRequired,
  setLiaison: propTypes.func.isRequired,
  send: propTypes.func.isRequired,
  initProposals: propTypes.array.isRequired,
  username: propTypes.string.isRequired,
  canAssign: propTypes.bool.isRequired
}
export default LiaisonSelector
