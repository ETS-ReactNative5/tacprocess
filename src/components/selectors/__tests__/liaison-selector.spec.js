import React from "react"
import renderer from "react-test-renderer"
import { shallow, mount, render } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import LiaisonSelector from "../LiaisonSelector"

const proposals = [
  { proposalCode: 'Code-1', liaisonAstronomer: 'LA-1' },
  { proposalCode: 'Code-2', liaisonAstronomer: 'LA-2' },
  { proposalCode: 'Code-3', liaisonAstronomer: 'LA-3'}
]

const initProposals = [ { proposalCode: 'CODE', liaisonAstronomer: 'LIAISON' }, { proposalCode: 'Code-3', liaisonAstronomer: 'LA-3'} ]

const canAssign = [true, false]

const astronomers = [ { username: 'LA-1', name: 'LA-1' }, { username: 'LA-2', name: 'LA-2' }, { username: 'LA-3', name: 'LA-3' } ]

const username = 'myezasifiso'

const setLiaison = jest.fn()

// Checking if the LiaisonSelector Component renders correctly for different input
describe("LiaisonSelector Component", () => {
  //Enzyme method testing if it renders correctly with empty values of the proposals
  it("Render the LiaisonSelector Component having unpopulated props with no errors", () => {
    const rendered = mount(<LiaisonSelector
      proposal= {{}}
      initProposals= {[]}
      canAssign= {canAssign[1]}
      astronomers= {[]}
      username = {''}
      setLiaison= {setLiaison} />)
    expect(shallowToJson(rendered)).toMatchSnapshot()
  })
  //Enzyme method testing if it renders correctly with populated values of the proposals
  it("Render the LiaisonSelector Component having populated props with no errors", () => {
    const rendered = mount(<LiaisonSelector
      proposal= {proposals}
      initProposals= {initProposals}
      canAssign= {canAssign[0]}
      astronomers= {astronomers}
      username= {username}
      setLiaison= {setLiaison} />)
    expect(shallowToJson(rendered)).toMatchSnapshot()
  })
})

// Checking if the setLiaison function works properly in LiaisonSelector Component
describe("LiaisonSelector Component", () => {
  it("setLiaison function should be called with the correct proposalCode", () => {
    const wrapper = mount(<LiaisonSelector
      proposal= {proposals[0]}
      initProposals= {initProposals}
      canAssign= {canAssign[0]}
      astronomers= {astronomers}
      username= {username}
      setLiaison= {setLiaison} />)

    wrapper.find('select').simulate('change');
    //Expect the setLiaison function to be called once
    expect(setLiaison.mock.calls.length).toBe(1)
    //Expect the setLiaison function to be called with the correct username
    expect(setLiaison.mock.calls[0][0]).toBe(astronomers[0].username)
    //Expect the setLiaison function to be called with the correct proposalCode
    expect(setLiaison.mock.calls[0][1]).toBe(proposals[0].proposalCode)
  })
})

// Checking if the drop down exist only for ADMINISTRATOR in LiaisonSelector Component
describe("LiaisonSelector Component", () => {
  it("Drop down should exist if the user is an ADMINISTRATOR", () => {
    const wrapper = mount(<LiaisonSelector
      proposal= {proposals[1]}
      initProposals= {initProposals}
      canAssign= {canAssign[0]}
      astronomers= {astronomers}
      username= {username}
      setLiaison= {setLiaison} />)
      // Expect the Drop down to exists (select)
      expect(wrapper.find('.setLiaison').type()).toBe('select')
  })

  it("Drop down should not exist but checkbox should if user is not an ADMINISTRATOR", () => {
    const wrapper = mount(<LiaisonSelector
      proposal= {proposals[2]}
      initProposals= {initProposals}
      canAssign= {canAssign[1]}
      astronomers= {astronomers}
      username= {username}
      setLiaison= {setLiaison} />)
      // Expect the Checkbox to exist (input)
      expect(wrapper.find('.setLiaison').type()).toBe('input')
  })
})
