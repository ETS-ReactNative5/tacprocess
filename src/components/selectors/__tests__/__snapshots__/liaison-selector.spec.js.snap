// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiaisonSelector Component Render the LiaisonSelector Component having populated props with no errors 1`] = `
<LiaisonSelector
  astronomers={
    Array [
      Object {
        "name": "LA-1",
        "username": "LA-1",
      },
      Object {
        "name": "LA-2",
        "username": "LA-2",
      },
      Object {
        "name": "LA-3",
        "username": "LA-3",
      },
    ]
  }
  canAssign={true}
  initProposals={
    Array [
      Object {
        "liaisonAstronomer": "LIAISON",
        "proposalCode": "CODE",
      },
      Object {
        "liaisonAstronomer": "LA-3",
        "proposalCode": "Code-3",
      },
    ]
  }
  proposal={
    Array [
      Object {
        "liaisonAstronomer": "LA-1",
        "proposalCode": "Code-1",
      },
      Object {
        "liaisonAstronomer": "LA-2",
        "proposalCode": "Code-2",
      },
      Object {
        "liaisonAstronomer": "LA-3",
        "proposalCode": "Code-3",
      },
    ]
  }
  setLiaison={[Function]}
  username="myezasifiso"
>
  <td>
    <select
      className="setLiaison"
      defaultValue={null}
      name="selector"
      onChange={[Function]}
      style={
        Object {
          "color": "blue",
        }
      }
    >
      <option
        value="none"
      >
        none
      </option>
      <option
        key="LA-1"
        name="LA-1"
        value="LA-1"
      >
        LA-1
      </option>
      <option
        key="LA-2"
        name="LA-2"
        value="LA-2"
      >
        LA-2
      </option>
      <option
        key="LA-3"
        name="LA-3"
        value="LA-3"
      >
        LA-3
      </option>
    </select>
  </td>
</LiaisonSelector>
`;

exports[`LiaisonSelector Component Render the LiaisonSelector Component having unpopulated props with no errors 1`] = `
<LiaisonSelector
  astronomers={Array []}
  canAssign={false}
  initProposals={Array []}
  proposal={Object {}}
  setLiaison={[Function]}
  username=""
>
  <td>
    <div>
      <input
        className="setLiaison"
        disabled={false}
        onChange={[Function]}
        type="checkbox"
        value=""
      />
      <a
        style={
          Object {
            "color": "red",
          }
        }
      >
        Select
      </a>
    </div>
  </td>
</LiaisonSelector>
`;
